Widgets_CDAT/DictWidget/dict.py

```diff
@@ -0,0 +1,219 @@
+from PySide.QtCore import *
+from PySide.QtGui import *
+from functools import partial
+
+class InputChecker(QValidator):
+
+    inputInvalid = Signal()
+    correctInput = Signal()
+
+    def __init__(self, index):
+        super(InputChecker, self).__init__()
+        self.texts = []
+        self.index = index
+
+    def validate(self, input, pos):
+        if input in self.texts:
+            if self.texts.index(input) != self.index:
+                self.inputInvalid.emit()
+                return QValidator.Intermediate
+
+        self.correctInput.emit()
+        return QValidator.Acceptable
+
+    def setIndex(self, index):
+        self.index = index
+
+
+class DictEditor(QWidget):
+    textEdited = Signal(dict)
+
+    def __init__(self):
+        super(DictEditor, self).__init__()
+        self.valid_keys = None
+        self.keys = []
+        self.values = []
+        self.textEdits = []
+        self.validators = []
+        self.row_count = 0
+        self.rows = QVBoxLayout()
+        add_button = QPushButton()
+        add_button.setText("New Line")
+        add_button.clicked.connect(self.insertRow)
+
+        self.rows.addWidget(add_button)
+        self.setLayout(self.rows)
+
+    def colorInvalid(self, line_edit):
+            line_edit.setStyleSheet("color: rgb(255, 0, 0);")
+
+    def colorValid(self, line_edit):
+            line_edit.setStyleSheet("color: rgb(0, 0, 0);")
+
+
+    # Update Combo Boxes
+    def updateCBoxes(self, cur_box):
+        cur_text = cur_box.itemText(cur_box.currentIndex())
+        selectedItems = []
+        for box in self.cBoxes:
+            check_text = box.itemText(box.currentIndex())
+
+            selectedItems.append(check_text)
+
+            # check if boxes match and set to blank if so
+            if box != cur_box:
+                if check_text == cur_text:
+                    box.setCurrentIndex(0)
+
+        return selectedItems
+
+
+    # updates dict based on lineedit values
+    def updateEditTexts(self):
+        texts = []
+        for item in self.textEdits:
+            texts.append(item.text())
+        for i in self.validators:
+            i.texts = self.keys
+
+
+    # update lists and selectedItems
+    def updateLists(self, left, right, index=None):
+        if self.valid_keys:
+            selectedItems = self.updateCBoxes(left)
+            self.keys = selectedItems
+            index = self.keys.index(left.itemText(left.currentIndex()))
+            self.values[index] = right.text()
+
+
+        else:
+            key = left.text()
+            value = right.text()
+            index = self.textEdits.index(left)
+            self.keys[index] = key
+            self.values[index] = value
+            self.updateEditTexts()
+
+        self.textEdited.emit(dict(zip(self.keys, self.values)))
```

Comment:
How about you iterate across self.rows and check to see if the key editors have valid input? If you do, you can emit, otherwise, don't.